sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java

```diff
@@ -207,97 +200,181 @@ public void doCollect() {
         visitorEnabled = fieldsVisitor.required();
         SearchContext.setCurrent(searchContext);
         searchContext.searcher().inStage(ContextIndexSearcher.Stage.MAIN_QUERY);
-        Query query = searchContext.query();
+        doCollect(null);
+    }
 
+    private void doCollect(OrderedCollectContext collectContext) {
         try {
-            assert query != null : "query must not be null";
-            if(orderBy != null) {
-                searchWithOrderBy(query);
+            if (collectContext == null) {
+                Query query = searchContext.query();
+                assert query != null : "query must not be null";
+                if(orderBy != null) {
+                    searchWithOrderBy(query);
+                } else {
+                    searchContext.searcher().search(query, this);
+                }
             } else {
-                searchContext.searcher().search(query, this);
+                searchAndCollect(collectContext);
+            }
+            if (!paused.get()) {
+                downstream.finish();
             }
-            downstream.finish();
         } catch (CollectionFinishedEarlyException e) {
+            paused.set(false);
             downstream.finish();
         } catch (Throwable e) {
+            paused.set(false);
             searchContext.close();
             downstream.fail(e);
         } finally {
-            if (rowCount == 0) {
-                searchContext.close();
+            if (!paused.get()) {
+                if (rowCount == 0) {
+                    searchContext.close();
+                }
+                searchContext.searcher().finishStage(ContextIndexSearcher.Stage.MAIN_QUERY);
+                assert SearchContext.current() == searchContext;
+                searchContext.clearReleasables(SearchContext.Lifetime.PHASE);
+                SearchContext.removeCurrent();
             }
-            searchContext.searcher().finishStage(ContextIndexSearcher.Stage.MAIN_QUERY);
-            assert SearchContext.current() == searchContext;
-            searchContext.clearReleasables(SearchContext.Lifetime.PHASE);
-            SearchContext.removeCurrent();
         }
     }
 
     @Override
     public void kill() {
         killed = true;
+        if (paused.get()) {
+            doCollect(collectContext);
```

Comment:
problem is that i want the catch()  { .. } finally {...} blocks from doCollect() to be called